resource-registry/src/Control/ResourceRegistry.hs

```diff
@@ -1528,3 +1539,49 @@ instance
   NoThunks (Bimap k v)
   where
   wNoThunks ctxt = noThunksInKeysAndValues ctxt . Bimap.toList
+
+-- | Transferring individual resources between registries is risky because a
+-- later resource in the origin registry might depend on a resource that we are
+-- trying to transfer. However, transferring whole registries is fine.
+--
+-- This will move all the resources from the origin registry to the destination
+-- registry, and will close the origin registry.
+ingestRegistry ::
+  (MonadSTM m, MonadMask m, MonadThread m, HasCallStack) =>
+  ResourceRegistry m ->
+  ResourceRegistry m ->
+  m [ResourceKey m]
+ingestRegistry fromReg toReg = do
+  context <- captureContext
+
+  -- The calling thread is known to the origin registry
+  ensureKnownThread fromReg context
+
+  -- Alloc all the needed keys
+  mKeys <- updateState toReg . allocNKeys =<< countResources fromReg
+
+  case mKeys of
+    -- If the destination registry is closed, throw
+    Left closed -> throwRegistryClosed toReg context closed
+    Right keys -> mask_ $ do
+      -- Get the resources out of the origin registry and empty it
+      regState <- atomically $ swapTVar (registryState fromReg) initState
+
+      mapM_
+        ( \(k, res) -> do
+            -- Insert the resources into the new registry
+            inserted <- updateState toReg (insertResource k res{resourceTransferredContext = Just context})
+
+            case inserted of
+              -- If the destination registry is closed, throw
+              Left closed -> do
+                let Release rel = resourceRelease res
+                void rel
+                throwRegistryClosed toReg context closed
```

Comment:
Don't know the library at all, just curious if you can get the underlying registry state (i presume Data.Map) earlier in the prelude of the function. So you perform these checks of whether registries are closed at the beginning and get the underlying states to work with purely.